## Airplane Passenger Satisfaction Veri seti Analizi Raporu
## Hazırlayan : Doğa Deniz
## Kullanılan Veri Seti : Airline Passenger Satisfaction / Kaggle

### Veri Seti Açıklaması 
* Bu veri seti, havayolu yolcularına ait çeşitli bilgileri ve uçuş deneyimlerine dair memnuniyet düzeylerini içermektedir.
*  Veri setinde:
Cinsiyet, yaş ve müşteri tipi gibi demografik bilgiler,
Uçuş mesafesi, seyahat amacı (kişisel/iş) ve uçuş sınıfı (ekonomi, business vb.) gibi uçuşa dair detaylar,
Yiyecek-içecek, koltuk konforu, online check-in, uçağın temizliği, eğlence sistemi, biniş kapısı konumu gibi hizmetlere verilen 0-5 arası memnuniyet puanları, Kalkış ve varış gecikme süreleri gibi operasyonel bilgiler
bulunmaktadır.
* Veri setindeki hedef değişken (target feature) Satisfaction (Memnuniyet) olup, yolcunun uçuş deneyiminden genel olarak memnun olup olmadığını belirtmektedir.

### Libraries:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Uploading Dataset

In [2]:
df = pd.read_csv("flight_train.csv")
df

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0                0   70172    Male     Loyal Customer   13  Personal Travel   
1                1    5047    Male  disloyal Customer   25  Business travel   
2                2  110028  Female     Loyal Customer   26  Business travel   
3                3   24026  Female     Loyal Customer   25  Business travel   
4                4  119299    Male     Loyal Customer   61  Business travel   
...            ...     ...     ...                ...  ...              ...   
103899      103899   94171  Female  disloyal Customer   23  Business travel   
103900      103900   73097    Male     Loyal Customer   49  Business travel   
103901      103901   68825    Male  disloyal Customer   30  Business travel   
103902      103902   54173  Female  disloyal Customer   22  Business travel   
103903      103903   62567    Male     Loyal Customer   27  Business travel   

           Class  Flight Distance  Inflight wifi service  \
0       Eco Plus              460                      3   
1       Business              235                      3   
2       Business             1142                      2   
3       Business              562                      2   
4       Business              214                      3   
...          ...              ...                    ...   
103899       Eco              192                      2   
103900  Business             2347                      4   
103901  Business             1995                      1   
103902       Eco             1000                      1   
103903  Business             1723                      1   

        Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                       4  ...                       5   
1                                       2  ...                       1   
2                                       2  ...                       5   
3                                       5  ...                       2   
4                                       3  ...                       3   
...                                   ...  ...                     ...   
103899                                  1  ...                       2   
103900                                  4  ...                       5   
103901                                  1  ...                       4   
103902                                  1  ...                       1   
103903                                  3  ...                       1   

        On-board service  Leg room service  Baggage handling  Checkin service  \
0                      4                 3                 4                4   
1                      1                 5                 3                1   
2                      4                 3                 4                4   
3                      2                 5                 3                1   
4                      3                 4                 4                3   
...                  ...               ...               ...              ...   
103899                 3                 1                 4                2   
103900                 5                 5                 5                5   
103901                 3                 2                 4                5   
103902                 4                 5                 1                5   
103903                 1                 1                 4                4   

        Inflight service  Cleanliness  Departure Delay in Minutes  \
0                      5            5                          25   
1                      4            1                           1   
2                      4            5                           0   
3                      4            2                          11   
4                      3            3                           0   
...                  ...          ...                         

In [3]:
df.shape
# 25 variables
# 103904 rows

(103904, 25)

Veri seti toplamda 103.904 gözlem ve birçok sayısal değişken içermektedir.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

### Statistical analysis for numerical variables

In [5]:
df.describe()

Unnamed: 0             id            Age  Flight Distance  \
count  103904.000000  103904.000000  103904.000000    103904.000000   
mean    51951.500000   64924.210502      39.379706      1189.448375   
std     29994.645522   37463.812252      15.114964       997.147281   
min         0.000000       1.000000       7.000000        31.000000   
25%     25975.750000   32533.750000      27.000000       414.000000   
50%     51951.500000   64856.500000      40.000000       843.000000   
75%     77927.250000   97368.250000      51.000000      1743.000000   
max    103903.000000  129880.000000      85.000000      4983.000000   

       Inflight wifi service  Departure/Arrival time convenient  \
count          103904.000000                      103904.000000   
mean                2.729683                           3.060296   
std                 1.327829                           1.525075   
min                 0.000000                           0.000000   
25%                 2.000000                           2.000000   
50%                 3.000000                           3.000000   
75%                 4.000000                           4.000000   
max                 5.000000                           5.000000   

       Ease of Online booking  Gate location  Food and drink  Online boarding  \
count           103904.000000  103904.000000   103904.000000    103904.000000   
mean                 2.756901       2.976883        3.202129         3.250375   
std                  1.398929       1.277621        1.329533         1.349509   
min                  0.000000       0.000000        0.000000         0.000000   
25%                  2.000000       2.000000        2.000000         2.000000   
50%                  3.000000       3.000000        3.000000         3.000000   
75%                  4.000000       4.000000        4.000000         4.000000   
max                  5.000000       5.000000        5.000000         5.000000   

        Seat comfort  Inflight entertainment  On-board service  \
count  103904.000000           103904.000000     103904.000000   
mean        3.439396                3.358158          3.382363   
std         1.319088                1.332991          1.288354   
min         0.000000                0.000000          0.000000   
25%         2.000000                2.000000          2.000000   
50%         4.000000                4.000000          4.000000   
75%         5.000000                4.000000          4.000000   
max         5.000000                5.000000          5.000000   

       Leg room service  Baggage handling  Checkin service  Inflight service  \
count     103904.000000     103904.000000    103904.000000     103904.000000   
mean           3.351055          3.631833         3.304290          3.640428   
std            1.315605          1.180903         1.265396          1.175663   
min            0.000000          1.000000         0.000000          0.000000   
25%            2.000000          3.000000         3.000000          3.000000   
50%            4.000000          4.000000         3.000000          4.000000   
75%            4.000000          5.000000         4.000000          5.000000   
max            5.000000          5.000000         5.000000          5.000000   

         Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  
count  103904.000000               103904.000000             103594.000000  
mean        3.286351                   14.815618                 15.178678  
std         1.312273                   38.230901                 38.698682  
min         0.000000                    0.000000                  0.000000  
25%         2.000000                    0.000000                  0.000000  
50%         3.000000                    0.000000                  0.000000  
75%         4.000000                   12.000000                 13.000000  
max         5.000000                 1592.000000               1584.000000

### AGE
* Ortalama yaş 39,38 olup, medyan değeri 40’tır. Bu durum yaş dağılımının oldukça simetrik olduğunu göstermektedir.
*  Standart sapma 15,11’dir; bu da yaş değerlerinin geniş bir aralıkta dağıldığını göstermektedir.
* Gözlemler 7 ile 85 yaş arasında değişmektedir.

### Flight Distance (Uçuş Mesafesi)
* Ortalama mesafe 1189,45 km, medyan ise 843 km’dir. Ortalama değerin medyandan büyük olması, değişkenin sağa çarpık (right-skewed) bir dağılıma sahip olduğunu göstermektedir.
* Standart sapma oldukça yüksek olup 997,15’tir. Bu durum uçuş mesafelerinin oldukça değişken olduğunu ve bazı ekstrem (çok uzun) uçuşların bulunduğunu göstermektedir.

### Departure Delay in Minutes (Kalkış Gecikmesi)
* Ortalama gecikme süresi 14,82 dakika, medyan ise 0’dır. Bu durum, gecikmelerin büyük bir kısmının kısa süreli veya hiç olmadığını göstermektedir.
* Standart sapma 38,23’tür. Maksimum gecikme süresi 1592 dakikadır.
* Gözlemlerin %75’i 0 ile 12 dakika arasında bir kalkış gecikmesine sahiptir

 ### Arrival Delay in Minutes (Varış Gecikmesi)
* Ortalama 15,18 dakika, medyan yine 0’dır. Kalkış gecikmesine benzer bir dağılım göstermektedir.
* Standart sapma 38,70 olup, maksimum gecikme 1584 dakikadır.
* Gözlemlerin %75’i 13 dakika veya altında varış gecikmesine sahiptir.

### Memnuniyet Skorları (Satisfaction Variables)
* Bu kategorideki değişkenler 0 ile 5 arasında değerlendirilmektedir. Ortalama skorlar genellikle 2,7 ila 3,6 arasında değişmektedir.
Standart sapmalar yaklaşık 1,3 düzeyindedir ve orta seviyede bir değişkenlik göstermektedir.

### Separate categorical and numeric columns

In [6]:
def grab_col_names(dataframe, cat_th=10, car_th=20):
    """
    grab_col_names for given dataframe

    :param dataframe:
    :param cat_th:
    :param car_th:
    :return:
    """

    cat_cols = [col for col in dataframe.columns if dataframe[col].dtypes == "O"]

    num_but_cat = [col for col in dataframe.columns if dataframe[col].nunique() < cat_th and
                   dataframe[col].dtypes != "O"]

    cat_but_car = [col for col in dataframe.columns if dataframe[col].nunique() > car_th and
                   dataframe[col].dtypes == "O"]

    cat_cols = cat_cols + num_but_cat
    cat_cols = [col for col in cat_cols if col not in cat_but_car]

    num_cols = [col for col in dataframe.columns if dataframe[col].dtypes != "O"]
    num_cols = [col for col in num_cols if col not in num_but_cat]

    print(f"Observations: {dataframe.shape[0]}")
    print(f"Variables: {dataframe.shape[1]}")
    print(f'cat_cols: {len(cat_cols)}')
    print(f'num_cols: {len(num_cols)}')
    print(f'cat_but_car: {len(cat_but_car)}')
    print(f'num_but_cat: {len(num_but_cat)}')

    

    return cat_cols, cat_but_car, num_cols, num_but_cat

cat_cols, cat_but_car, num_cols, num_but_cat = grab_col_names(df)

Observations: 103904
Variables: 25
cat_cols: 19
num_cols: 6
cat_but_car: 0
num_but_cat: 14


In [7]:
cat_cols

['Gender',
 'Customer Type',
 'Type of Travel',
 'Class',
 'satisfaction',
 'Inflight wifi service',
 'Departure/Arrival time convenient',
 'Ease of Online booking',
 'Gate location',
 'Food and drink',
 'Online boarding',
 'Seat comfort',
 'Inflight entertainment',
 'On-board service',
 'Leg room service',
 'Baggage handling',
 'Checkin service',
 'Inflight service',
 'Cleanliness']

In [8]:
num_but_cat

['Inflight wifi service',
 'Departure/Arrival time convenient',
 'Ease of Online booking',
 'Gate location',
 'Food and drink',
 'Online boarding',
 'Seat comfort',
 'Inflight entertainment',
 'On-board service',
 'Leg room service',
 'Baggage handling',
 'Checkin service',
 'Inflight service',
 'Cleanliness']

In [9]:
num_cols

['Unnamed: 0',
 'id',
 'Age',
 'Flight Distance',
 'Departure Delay in Minutes',
 'Arrival Delay in Minutes']

### Analysis of Categorical Columns


In [10]:

def cat_summary(dataframe, col_name, plot=False):
    print(pd.DataFrame({col_name: dataframe[col_name].value_counts(),
                        "Ratio": 100 * dataframe[col_name].value_counts() / len(dataframe)}))
    print("##########################################")
    if plot:
        sns.countplot(x=dataframe[col_name], data=dataframe)
        plt.show()

for col in cat_cols:
    cat_summary(df, col)

        Gender      Ratio
Gender                   
Female   52727  50.745881
Male     51177  49.254119
##########################################
                   Customer Type      Ratio
Customer Type                              
Loyal Customer             84923  81.732176
disloyal Customer          18981  18.267824
##########################################
                 Type of Travel      Ratio
Type of Travel                            
Business travel           71655  68.962696
Personal Travel           32249  31.037304
##########################################
          Class      Ratio
Class                     
Business  49665  47.798930
Eco       46745  44.988643
Eco Plus   7494   7.212427
##########################################
                         satisfaction      Ratio
satisfaction                                    
neutral or dissatisfied         58879  56.666731
satisfied                       45025  43.333269
##########################################
  

## Kategorik Değişkenlerin Dağılım Analizi
Bu bölümde, veri setinde yer alan kategorik değişkenlerin frekans ve oran analizleri sunulmuştur. Her bir değişken için baskın sınıflar belirlenmiş, dağılımlar açıklanmış ve veri dengesine yönelik genel bir değerlendirme yapılmıştır.

1. Gender (Cinsiyet)
Kadın yolcu oranı %50,75, erkek yolcu oranı ise %49,25’tir.
Cinsiyet değişkeni dengeli bir dağılıma sahiptir, modelleme açısından denge sorunu beklenmemektedir.

2. Customer Type (Müşteri Türü)
Yolcuların %81,73’ü "Loyal Customer" (sadık müşteri) iken, %18,27’si "Disloyal Customer" (sadık olmayan müşteri) grubundadır.
Bu dağılım, müşteri sadakatine göre dengesiz bir yapıya sahiptir. Model eğitiminde bu dengesizlik dikkate alınmalıdır.

3. Type of Travel (Seyahat Türü)
%68,96’sı "Business Travel" (iş seyahati), %31,04’ü ise "Personal Travel" (kişisel seyahat) olarak gerçekleşmiştir.
İş seyahatleri, veri setinde daha baskın bir rol oynamaktadır.

4. Class (Uçuş Sınıfı)
Business: %47,80
Eco: %44,99
Eco Plus: %7,21
Yolcuların çoğunluğu Business ve Economy sınıfında yer almaktadır. Eco Plus sınıfı oldukça düşük bir paya sahiptir.

5. Satisfaction (Memnuniyet Durumu)
%56,67’si “neutral or dissatisfied” (nötr veya memnun olmayan)
%43,33’ü “satisfied” (memnun)
Veri seti, memnun olmayan yolcular lehine bir ağırlık içermektedir.

## Skor Bazlı Kategorik Değişkenler
Skorlar genellikle 0–5 arasında değerlendirilmiştir. 0 değerleri genellikle "uygulanamaz" durumları temsil eder.
### 6. Inflight Wifi Service
En sık puanlar: 3 (%24,90) ve 2 (%24,86)
0 puan oranı %2,99’dur.
Orta düzeyde memnuniyet eğilimi gözlemlenmektedir.
### 7. Departure/Arrival Time Convenient
En sık puan: 4 (%24,59), ardından 5 (%21,56)
0 puan oranı %5,10’dur.
Bu hizmetin genel olarak olumlu karşılandığı görülmektedir.
### 8. Ease of Online Booking
Dağılım: Puan 3 (%23,53) ve 2 (%23,12) ile öne çıkmaktadır.
0 puan alanların oranı %4,32’dir.
### 9. Gate Location
En yüksek oran: Puan 3 (%27,50)
0 puan oranı %0,001 civarındadır. Neredeyse yok denecek kadar azdır.
### 10. Food and Drink
Puan 4 (%23,44) en yaygın tercihtir.
0 puan oranı oldukça düşük: %0,10
### 11. Online Boarding
En yüksek oran: Puan 4 (%29,61)
0 puan oranı %2,34
### 12. Seat Comfort
En yaygın puan: 4 (%30,57)
0 puan neredeyse yok: %0,001
### 13. Inflight Entertainment
Öne çıkan puan: 4 (%28,32) ve 5 (%24,27)
0 puan oranı %0,01
### 14. On-board Service
Puan 4 (%29,71) ve 5 (%22,76) öne çıkmakta.
0 puan oranı yok denecek kadar az: %0,003
### 15. Leg Room Service
Öne çıkan puanlar: 4 (%27,71) ve 5 (%23,74)
0 puan oranı %0,45
### 16.Baggage Handling
En yaygın puan: 4 (%35,98)
Genel olarak olumlu bir algı söz konusudur.
### 17. Check-in Service
Puan 4 ve 3 oranları benzer: %27,96 ve %27,38
0 puan oranı %0,001
### 18.Inflight Service
En yaygın puan: 4 (%36,52)
0 puan oranı yok denecek kadar düşük.
### 19. Cleanliness (Temizlik)
En yaygın puan: 4 (%26,16)
0 puan oranı %0,01 olup ihmal edilebilir düzeydedir.


## Missing Value Analysis

Arrival Delay in Minutes değişkeninde 310 adet eksik değer var

In [2]:
df.isnull().sum()

NameError: name 'df' is not defined

In [39]:
# eksik değerler ortalama ile doldurulacak
df['Arrival Delay in Minutes'].fillna(df['Arrival Delay in Minutes'].mean(),inplace=True)

C:\Users\90530\AppData\Local\Temp\ipykernel_60348\1738186449.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Arrival Delay in Minutes'].fillna(df['Arrival Delay in Minutes'].mean(),inplace=True)


In [40]:
df['Arrival Delay in Minutes'].isnull().sum()
# Başarılı

0

## Outlier Analysis for Numerical Columns

In [11]:
num_cols

['Unnamed: 0',
 'id',
 'Age',
 'Flight Distance',
 'Departure Delay in Minutes',
 'Arrival Delay in Minutes']

In [12]:
# gereksiz sütunları sil
num_cols = [col for col in num_cols if col not in ['Unnamed: 0','id']]
num_cols

['Age',
 'Flight Distance',
 'Departure Delay in Minutes',
 'Arrival Delay in Minutes']

In [29]:
def outlier_threshold(dataframe,col_name,q1=0.10,q3=0.90):
    Q1 = dataframe[col_name].quantile(q1)
    Q3 = dataframe[col_name].quantile(q3)

    IQR = Q3 - Q1
    lower_limit = Q1 - 1.5 * IQR
    upper_limit = Q3 + 1.5 * IQR

    return lower_limit,upper_limit
    

In [30]:
def check_outlier(dataframe,col_name):
    lower_limit,upper_limit =  outlier_threshold(dataframe,col_name)
    if dataframe[(dataframe[col_name] < lower_limit)| (dataframe[col_name] > upper_limit)].any(axis=None):
        return True
    else:
        return False

In [31]:
for col in num_cols:
    print(col,check_outlier(df,col))
    

Age False
Flight Distance False
Departure Delay in Minutes True
Arrival Delay in Minutes True


In [32]:

# aykırı değerlerler threshold value kullanılarak limit değerlere eşitlenir
def replace_with_thresholds(dataframe,col_name):
    dataframe[col_name] = dataframe[col_name].astype(float)
    low_limit, up_limit = outlier_threshold(dataframe,col_name)
    dataframe.loc[(dataframe[col_name] < low_limit),col_name] = low_limit
    dataframe.loc[(dataframe[col_name] > up_limit),col_name] = up_limit

In [33]:
for col in num_cols:
    replace_with_thresholds(df,col)

In [35]:
for col in num_cols:
    print(col,check_outlier(df,col))
# Başarılı 

Age False
Flight Distance False
Departure Delay in Minutes False
Arrival Delay in Minutes False



## Veri Temizleme ve Hazırlık (Data Cleaning & Preparation)
* Arrival Delay in Minutes sütununda eksik veriler bulunmuştur.Bu eksik veriler ortalama alma yöntemiyle ( df['Arrival Delay in Minutes'].mean() ) doldurulmuştur.
* Sayısal sütunlar için Aykırı Değer analizi yapılmıştır
* Veri setindeki sayısal değişkenlerde uç değerleri belirlemek için IQR (Interquartile Range / Çeyrekler Arası Aralık) yöntemini kullanıldı. Bu yöntemle:
1) Q1 (1. çeyrek) ve Q3 (3. çeyrek) değerlerini hesaplandı.Default olarak Q1= 0.25 ,Q3 = 0.75 kullanıldı
2) IQR = Q3 - Q1 formülüyle çeyrekler arası aralığı bulunuldu.
3) Uç değer sınırlarını aşağıdaki şekilde tanımlandı:
 * Alt sınır = Q1 - 1.5 × IQR
 * Üst sınır = Q3 + 1.5 × IQR
* Bu sınırların altında veya üstünde kalan değerleri uç değer (outlier) olarak kabul edildi.
* Gözlemlenen bazı değişkenlerde (örneğin Departure Delay in Minutes ve Arrival Delay in Minutes) ciddi sayıda uç değer vardı.
Uç değerleri tamamen silmek yerine, analizde sapmalara neden olmamaları için bu değerler belirlenen alt ve üst sınırlarla sınırlandı (capping).
Böylece hem veri kaybı yaşanmadı hem de uç değerlerin olumsuz etkisi azaltılmış oldu.
 
* Sütunları sayısal ve kategorik olarak ayırmak için grab_col_names fonksiyonu kullanıldı.